File: sdk/typescript/src/e2b/sandbox.ts
/** `e2b`'s `Sandbox`, backed by `@arcbox/sandbox`. */

import type {
  ConnectionOptions,
  Sandbox as ArcBoxSandbox,
  SandboxState,
} from "../index";
import { ArcBox, NotFoundError } from "../index";

import { Commands, Pty } from "./commands";
import { unsupported } from "./errors";
import { Filesystem } from "./filesystem";
import { Git } from "./git";

/** `e2b`'s default sandbox timeout. */
export const DEFAULT_SANDBOX_TIMEOUT_MS = 300000;

/**
 * Connection options.
 *
 * The E2B cloud knobs (`apiKey`, `domain`, `accessToken`, …) are
 * accepted and ignored — a local daemon authenticates by socket
 * permissions — so an app that reads `E2B_API_KEY` from its environment
 * keeps working unchanged. `connection` is the ArcBox escape hatch.
 */
export interface ConnectionOpts {
  apiKey?: string;
  accessToken?: string;
  domain?: string;
  apiUrl?: string;
  debug?: boolean;
  requestTimeoutMs?: number;
  headers?: Record<string, string>;
  /** ArcBox connection override; no `e2b` counterpart. */
  connection?: ConnectionOptions;
}

/** Options for {@link Sandbox.create}. */
export interface SandboxOpts extends ConnectionOpts {
  /** Template name; `''` and `'base'` both mean the built-in minimal one. */
  template?: string;
  /** Arbitrary metadata — ArcBox calls these labels. */
  metadata?: Record<string, string>;
  /** Environment for the sandbox's initial command. */
  envs?: Record<string, string>;
  /** Hard maximum lifetime. */
  timeoutMs?: number;
  /** `false` disables networking entirely. */
  allowInternetAccess?: boolean;
}

/** Options for {@link Sandbox.connect}. */
export type SandboxConnectOpts = ConnectionOpts & { timeoutMs?: number };

/** Options for {@link Sandbox.list}. */
export interface SandboxListOpts extends ConnectionOpts {
  query?: { metadata?: Record<string, string>; state?: SandboxState[] };
}

/** One row of a sandbox listing, in `e2b`'s shape. */
export interface SandboxInfo {
  sandboxId: string;
  templateId: string;
  name?: string;
  metadata: Record<string, string>;
  startedAt?: Date;
  endAt?: Date;
  state: SandboxState;
}

/**
 * What {@link Sandbox.list} returns: both of `e2b`'s listing shapes at
 * once. Awaiting it yields the whole listing (the daemon streams every
 * page already); calling {@link nextItems} pages e2b-style — one page,
 * carrying that same complete listing.
 */
export class SandboxPaginator implements PromiseLike<SandboxInfo[]> {
  readonly #all: Promise<SandboxInfo[]>;
  #consumed = false;

  constructor(all: Promise<SandboxInfo[]>) {
    this.#all = all;
  }

  /** Whether {@link nextItems} has a page left (exactly one, the whole listing). */
  get hasNext(): boolean {
    return !this.#consumed;
  }

  /** The next page: the complete listing first, `[]` from then on. */
  async nextItems(): Promise<SandboxInfo[]> {
    if (this.#consumed) {
      return [];
    }
    this.#consumed = true;
    return this.#all;
  }

  // eslint-disable-next-line sukka/unicorn/no-thenable -- deliberate: e2b v1 code awaits list() while v2 code holds the paginator; PromiseLike is the one shape that serves both
  then<TResult1 = SandboxInfo[], TResult2 = never>(
    onfulfilled?:
      | ((value: SandboxInfo[]) => TResult1 | PromiseLike<TResult1>)
      | null,
    onrejected?: ((reason: unknown) => TResult2 | PromiseLike<TResult2>) | null,
  ): PromiseLike<TResult1 | TResult2> {
    // eslint-disable-next-line promise/prefer-catch -- not a consumer call: this IS the PromiseLike implementation, forwarding both callbacks
    return this.#all.then(onfulfilled, onrejected);
  }
}

/**
 * `e2b`'s cloud sandbox, served by the local ArcBox daemon.
 *
 * The shape follows `e2b` deliberately, including the parts that read
 * oddly here: `sandboxId` rather than `id`, milliseconds everywhere,
 * `metadata` rather than `labels`, and `run()` throwing on a non-zero
 * exit. Divergences are documented on the members that carry them —
 * `getHost` most of all.
 */
export class Sandbox {
  /** Unique identifier of the sandbox. */
  readonly sandboxId: string;
  /** Interact with the sandbox filesystem. */
  readonly files: Filesystem;
  /** Run commands in the sandbox. */
  readonly commands: Commands;
  /** Drive pseudo-terminals in the sandbox. */
  readonly pty: Pty;
  /** Run git operations in the sandbox. */
  readonly git: Git;

  readonly #client: ArcBox;
  readonly #sandbox: ArcBoxSandbox;
  /** Host ports resolved by {@link exposePort}, keyed by sandbox port. */
  readonly #hosts = new Map<number, string>();

  constructor(client: ArcBox, sandbox: ArcBoxSandbox) {
    this.#client = client;
    this.#sandbox = sandbox;
    this.sandboxId = sandbox.id;
    this.files = new Filesystem(sandbox.files);
    this.commands = new Commands(sandbox.commands);
    this.pty = new Pty(sandbox.commands);
    this.git = new Git(this.commands);
  }

  /** Create a sandbox from a template. */
  static async create(this: void, opts?: SandboxOpts): Promise<Sandbox>;
  static async create(
    this: void,
    template: string,
    opts?: SandboxOpts,
  ): Promise<Sandbox>;
  static async create(
    this: void,
    templateOrOpts?: string | SandboxOpts,
    maybeOpts?: SandboxOpts,
  ): Promise<Sandbox> {
    const opts =
      (typeof templateOrOpts === "string" ? maybeOpts : templateOrOpts) ?? {};
    const template =
      typeof templateOrOpts === "string" ? templateOrOpts : opts.template;
    const client = new ArcBox(opts.connection);
    const sandbox = await client.create(normalizeTemplate(template), {
      ttlMs: opts.timeoutMs ?? DEFAULT_SANDBOX_TIMEOUT_MS,
      ...(opts.metadata !== undefined && { labels: opts.metadata }),
      ...(opts.envs !== undefined && { env: opts.envs }),
      ...(opts.allowInternetAccess !== undefined && {
        network: opts.allowInternetAccess,
      }),
    });
    return new Sandbox(client, sandbox);
  }

  /** Attach to an existing sandbox, resuming it when paused. */
  static async connect(
    this: void,
    sandboxId: string,
    opts: SandboxConnectOpts = {},
  ): Promise<Sandbox> {
    const client = new ArcBox(opts.connection);
    const sandbox = await client.connect(sandboxId);
    const handle = new Sandbox(client, sandbox);
    if (opts.timeoutMs !== undefined) {
      await handle.setTimeout(opts.timeoutMs);
    }
    return handle;
  }

  /**
   * List sandboxes, as a {@link SandboxPaginator} that is also
   * awaitable: `await Sandbox.list()` yields the whole listing, and
   * `Sandbox.list().nextItems()` pages e2b-style (one page carrying
   * that same complete listing — the daemon streams every page
   * already).
   */
  static list(this: void, opts: SandboxListOpts = {}): SandboxPaginator {
    const collect = async (): Promise<SandboxInfo[]> => {
      const client = new ArcBox(opts.connection);
      const wanted = opts.query?.state;
      const rows: SandboxInfo[] = [];
      for await (const summary of client.list({
        ...(opts.query?.metadata !== undefined && {
          labels: opts.query.metadata,
        }),
      })) {
        if (wanted !== undefined && !wanted.includes(summary.state)) {
          continue;
        }
        const row: SandboxInfo = {
          sandboxId: summary.id,
          templateId: "",
          metadata: summary.labels,
          state: summary.state,
        };
        if (summary.createdAt !== undefined) {
          row.startedAt = summary.createdAt;
        }
        rows.push(row);
      }
      return rows;
    };
    return new SandboxPaginator(collect());
  }

  /** Kill a sandbox by id. Resolves `false` when it was already gone. */
  static async kill(
    this: void,
    sandboxId: string,
    opts: ConnectionOpts = {},
  ): Promise<boolean> {
    try {
      await (await new ArcBox(opts.connection).connect(sandboxId)).kill();
      return true;
    } catch (error) {
      if (error instanceof NotFoundError) {
        return false;
      }
      throw error;
    }
  }

  /** Extend a sandbox's lifetime by id. */
  static async setTimeout(
    this: void,
    sandboxId: string,
    timeoutMs: number,
    opts: ConnectionOpts = {},
  ): Promise<void> {
    const sandbox = await new ArcBox(opts.connection).connect(sandboxId);
    await sandbox.setLifecycle({ ttlMs: timeoutMs });
  }

  /**
   * Whether the sandbox is running.
   *
   * `false` covers a missing sandbox too, matching `e2b`, where a killed
   * sandbox reports not-running rather than raising.
   */
  async isRunning(): Promise<boolean> {
    try {
      const info = await this.#sandbox.info();
      return info.state === "ready" || info.state === "running";
    } catch (error) {
      if (error instanceof NotFoundError) {
        return false;
      }
      throw error;
    }
  }

  /**
   * Re-arm the sandbox's lifetime, measured from now.
   *
   * `e2b` only ever extends; ArcBox replaces, so a shorter value here
   * genuinely shortens the sandbox's life.
   */
  async setTimeout(timeoutMs: number): Promise<void> {
    await this.#sandbox.setLifecycle({ ttlMs: timeoutMs });
  }

  /** Kill the sandbox. Resolves `false` when it was already gone. */
  async kill(): Promise<boolean> {
    try {
      await this.#sandbox.kill();
      return true;
    } catch (error) {
      if (error instanceof NotFoundError) {
        return false;
      }
      throw error;
    }
  }

  /** Pause the sandbox, checkpointing it under the same id. */
  async pause(): Promise<boolean> {
    await this.#sandbox.pause();
    return true;
  }

  /** Deprecated `e2b` alias for {@link pause}. */
  async betaPause(): Promise<boolean> {
    return this.pause();
  }

  /**
   * Resume the sandbox and return this handle.
   *
   * Routed through the entry point's connect, which settles a pausing
   * checkpoint, resumes a paused one, and waits out a starting boot.
   */
  async connect(): Promise<this> {
    await this.#client.connect(this.sandboxId);
    return this;
  }

  /** Current state of the sandbox, in `e2b`'s shape. */
  async getInfo(): Promise<SandboxInfo> {
    const info = await this.#sandbox.info();
    const out: SandboxInfo = {
      sandboxId: info.id,
      templateId: info.template,
      metadata: info.labels,
      state: info.state,
    };
    if (info.createdAt !== undefined) {
      out.startedAt = info.createdAt;
    }
    if (info.ttlDeadline !== undefined) {
      out.endAt = info.ttlDeadline;
    }
    return out;
  }

  /**
   * Publish a sandbox port onto the host and return its `host:port`.
   *
   * The one real divergence from `e2b`, which serves every sandbox port
   * through an edge proxy at `{port}-{id}.e2b.app` and so can answer
   * synchronously without being told. ArcBox forwards a port only once
   * it is asked to, which is what this awaits.
   */
  async exposePort(port: number): Promise<string> {
    const exposed = await this.#sandbox.ports.expose(port);
    const host = `127.0.0.1:${String(exposed.hostPort)}`;
    this.#hosts.set(port, host);
    return host;
  }

  /** Withdraw a port published by {@link exposePort}. */
  async unexposePort(port: number): Promise<void> {
    await this.#sandbox.ports.unexpose(port);
    this.#hosts.delete(port);
  }

  /**
   * `host:port` for a port already published by {@link exposePort}.
   *
   * Synchronous, like `e2b`'s — but a port this handle has not exposed
   * throws instead of returning a plausible address that nothing is
   * listening on.
   */
  getHost(port: number): string {
    const host = this.#hosts.get(port);
    if (host === undefined) {
      unsupported(
        "getHost",
        `port ${String(port)} is not published — await sandbox.exposePort(${String(port)}) first, since ArcBox forwards ports on request rather than through an edge proxy`,
      );
    }
    return host;
  }

  /** Unsupported: cloud snapshots that outlive their sandbox. */
  createSnapshot(): never {
    unsupported(
      "createSnapshot",
      "ArcBox snapshots are local; use @arcbox/sandbox's checkpoint()/restore() directly",
    );
  }

  /** Unsupported: forking needs the cloud's copy-on-write sandbox pool. */
  fork(): never {
    unsupported("fork", "no local equivalent");
  }

  /** Unsupported: per-sandbox metrics are a cloud service. */
  getMetrics(): never {
    unsupported("getMetrics", "no local equivalent");
  }

  /** Unsupported: signed URLs are served by the E2B edge. */
  uploadUrl(): never {
    unsupported("uploadUrl", "use files.write instead");
  }

  /** Unsupported: signed URLs are served by the E2B edge. */
  downloadUrl(): never {
    unsupported("downloadUrl", "use files.read instead");
  }

  /** Unsupported: the MCP gateway is a cloud service. */
  getMcpUrl(): never {
    unsupported("getMcpUrl", "no local equivalent");
  }
}

/**
 * `e2b`'s default template is `'base'`; ArcBox spells its built-in
 * minimal template `''`, so the two names converge here rather than
 * failing to resolve a template nobody published.
 */
function normalizeTemplate(template: string | undefined): string {
  return template === undefined || template === "base" ? "" : template;
}

File: sdk/python/src/arcbox/e2b/_async/sandbox.py
"""``e2b``'s ``Sandbox``, backed by the arcbox SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING

from arcbox import AsyncArcBox
from arcbox.e2b._types import (
    DEFAULT_SANDBOX_TIMEOUT_MS,
    SandboxInfo,
    normalize_template,
    seconds,
)
from arcbox.e2b.errors import unsupported
from arcbox.errors import NotFoundError

from .commands import AsyncCommands, AsyncPty
from .filesystem import AsyncFilesystem
from .git import AsyncGit

if TYPE_CHECKING:
    from collections.abc import Mapping
    from types import TracebackType

    from arcbox import AsyncSandbox as ArcBoxSandbox
    from arcbox import Connection


class AsyncSandbox:
    """``e2b``'s cloud sandbox, served by the local ArcBox daemon.

    The shape follows ``e2b`` deliberately, including the parts that
    read oddly here: ``sandbox_id`` rather than ``id``, milliseconds
    everywhere, ``metadata`` rather than ``labels``, and ``run()``
    raising on a non-zero exit. Divergences are documented on the
    members that carry them — :meth:`get_host` most of all.

    The E2B cloud arguments (``api_key``, ``domain``, ...) are accepted
    and ignored: a local daemon authenticates by socket permissions, so
    an app that reads ``E2B_API_KEY`` from its environment keeps
    working. ``connection`` is the arcbox escape hatch."""

    def __init__(self, sandbox: ArcBoxSandbox) -> None:
        self._sandbox = sandbox
        #: Unique identifier of the sandbox.
        self.sandbox_id = sandbox.id
        #: Interact with the sandbox filesystem.
        self.files = AsyncFilesystem(sandbox.files)
        #: Run commands in the sandbox.
        self.commands = AsyncCommands(sandbox.commands)
        #: Drive pseudo-terminals in the sandbox.
        self.pty = AsyncPty(sandbox.commands)
        #: Run git operations in the sandbox.
        self.git = AsyncGit(self.commands)
        # Host ports resolved by expose_port, keyed by sandbox port.
        self._hosts: dict[int, str] = {}

    @classmethod
    async def create(
        cls,
        template: str | None = None,
        *,
        metadata: Mapping[str, str] | None = None,
        envs: Mapping[str, str] | None = None,
        timeout_ms: float = DEFAULT_SANDBOX_TIMEOUT_MS,
        allow_internet_access: bool | None = None,
        connection: Connection | None = None,
        **_cloud: object,
    ) -> AsyncSandbox:
        """Create a sandbox from a template."""
        sandbox = await AsyncArcBox(connection).create(
            normalize_template(template),
            ttl=seconds(timeout_ms),
            labels=metadata,
            env=envs,
            network=allow_internet_access,
        )
        return cls(sandbox)

    @classmethod
    async def connect(
        cls,
        sandbox_id: str,
        *,
        timeout_ms: float | None = None,
        connection: Connection | None = None,
        **_cloud: object,
    ) -> AsyncSandbox:
        """Attach to an existing sandbox, resuming it when paused."""
        handle = cls(await AsyncArcBox(connection).connect(sandbox_id))
        if timeout_ms is not None:
            await handle.set_timeout(timeout_ms)
        return handle

    @classmethod
    async def list(
        cls,
        *,
        metadata: Mapping[str, str] | None = None,
        state: list[str] | None = None,
        connection: Connection | None = None,
        **_cloud: object,
    ) -> list[SandboxInfo]:
        """List sandboxes.

        ``e2b`` returns a paginator; the arcbox SDK streams every page
        already, so this returns the whole listing."""
        rows: list[SandboxInfo] = []
        async with AsyncArcBox(connection) as client:
            async for summary in client.list(labels=metadata):
                if state is not None and summary.state not in state:
                    continue
                rows.append(
                    SandboxInfo(
                        sandbox_id=summary.id,
                        template_id="",
                        metadata=summary.labels,
                        state=summary.state,
                        started_at=summary.created_at,
                    )
                )
        return rows

    async def is_running(self) -> bool:
        """Whether the sandbox is running.

        ``False`` covers a missing sandbox too, matching ``e2b``, where
        a killed sandbox reports not-running rather than raising."""
        try:
            info = await self._sandbox.info()
        except NotFoundError:
            return False
        return info.state in {"ready", "running"}

    async def set_timeout(self, timeout_ms: float) -> None:
        """Re-arm the sandbox's lifetime, measured from now.

        ``e2b`` only ever extends; arcbox replaces, so a shorter value
        here genuinely shortens the sandbox's life."""
        await self._sandbox.set_lifecycle(ttl=seconds(timeout_ms))

    async def kill(self) -> bool:
        """Kill the sandbox. ``False`` when it was already gone.

        ``e2b`` also exposes a ``Sandbox.kill(sandbox_id)`` class form;
        Python cannot carry both under one name, so reach a sandbox you
        do not hold through ``Sandbox.connect(sandbox_id)`` first."""
        try:
            await self._sandbox.kill()
        except NotFoundError:
            return False
        return True

    async def pause(self) -> bool:
        """Pause the sandbox, checkpointing it under the same id."""
        await self._sandbox.pause()
        return True

    async def beta_pause(self) -> bool:
        """Deprecated ``e2b`` alias for :meth:`pause`."""
        return await self.pause()

    async def get_info(self) -> SandboxInfo:
        """Current state of the sandbox, in ``e2b``'s shape."""
        info = await self._sandbox.info()
        return SandboxInfo(
            sandbox_id=info.id,
            template_id=info.template,
            metadata=info.labels,
            state=info.state,
            started_at=info.created_at,
            end_at=info.ttl_deadline,
        )

    async def expose_port(self, port: int) -> str:
        """Publish a sandbox port onto the host and return ``host:port``.

        The one real divergence from ``e2b``, which serves every sandbox
        port through an edge proxy at ``{port}-{id}.e2b.app`` and so can
        answer synchronously without being told. ArcBox forwards a port
        only once it is asked to, which is what this awaits."""
        exposed = await self._sandbox.ports.expose(port)
        host = f"127.0.0.1:{exposed.host_port}"
        self._hosts[port] = host
        return host

    async def unexpose_port(self, port: int) -> None:
        """Withdraw a port published by :meth:`expose_port`."""
        await self._sandbox.ports.unexpose(port)
        self._hosts.pop(port, None)

    def get_host(self, port: int) -> str:
        """``host:port`` for a port already published by :meth:`expose_port`.

        Synchronous, like ``e2b``'s — but a port this handle has not
        exposed raises instead of returning a plausible address that
        nothing is listening on."""
        host = self._hosts.get(port)
        if host is None:
            unsupported(
                "get_host",
                f"port {port} is not published — await sandbox.expose_port({port}) "
                "first, since ArcBox forwards ports on request rather than through "
                "an edge proxy",
            )
        return host

    def create_snapshot(self, *args: object, **kwargs: object) -> None:
        """Unsupported: cloud snapshots that outlive their sandbox."""
        del args, kwargs
        unsupported(
            "create_snapshot",
            "ArcBox snapshots are local; use arcbox's checkpoint()/restore() directly",
        )

    def fork(self, *args: object, **kwargs: object) -> None:
        """Unsupported: forking needs the cloud's copy-on-write pool."""
        del args, kwargs
        unsupported("fork", "no local equivalent")

    def get_metrics(self, *args: object, **kwargs: object) -> None:
        """Unsupported: per-sandbox metrics are a cloud service."""
        del args, kwargs
        unsupported("get_metrics", "no local equivalent")

    def upload_url(self, *args: object, **kwargs: object) -> None:
        """Unsupported: signed URLs are served by the E2B edge."""
        del args, kwargs
        unsupported("upload_url", "use files.write instead")

    def download_url(self, *args: object, **kwargs: object) -> None:
        """Unsupported: signed URLs are served by the E2B edge."""
        del args, kwargs
        unsupported("download_url", "use files.read instead")

    def get_mcp_url(self, *args: object, **kwargs: object) -> None:
        """Unsupported: the MCP gateway is a cloud service."""
        del args, kwargs
        unsupported("get_mcp_url", "no local equivalent")

    async def __aenter__(self) -> AsyncSandbox:
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        await self._sandbox.__aexit__(exc_type, exc, tb)
